### ISSUE TITLE
[FEATURE] Support query the unified Fileset Data Source in Spark

### ISSUE THREAD

**Issue by xloya** [apache/gravitino#5226]
### Describe the feature

In the current integration of Fileset and Spark, we can already read / write the data under Fileset in the form of gvfs path which file format supported by Spark:
```
// Dataframe API
val df = sparkSession
  .read
  .parquet("gvfs://fileset/fileset_catalog/tmp/test_fileset/date=20241021")

df.write
  .mode("overwrite")
  .parquet("gvfs://fileset/fileset_catalog/tmp/test_fileset/date=20241022")

// SQL API
SELECT * FROM `parquet`.`gvfs://fileset/fileset_catalog/tmp/test_fileset` WHERE date = 20241022

INSERT OVERWRITE DIRECTORY 'gvfs://fileset/fileset_catalog/tmp/test_fileset/date=20241023' USING parquet OPTIONS (col1 'name', col2 'age', col3 'test') SELECT * FROM test_table;
```
In fact, Databricks also handles Volume in this way: https://docs.databricks.com/en/sql/language-manual/sql-ref-volumes.html.

### Motivation

But there is two problems here:
1. If users want to read a specific file format that is not officially supported by Spark under Fileset through SQL, there is currently no way.
2. If the file format does not support schema inference, additional schema metadata is required.

### Describe the solution

My idea is that we should consider providing a unified Fileset data source which extends the Spark FileDatasourceV2 interface. For the file formats currently supported by Spark, we can reuse the relevant logic (such as parquet/csv/json, etc.); for user-defined file formats, users can extend and implement them themselves (such as sequence file, tfrecord, etc.). 
Then users can access the Fileset data source in a unified manner through the following forms. The specific file format and schema are obtained and routed through the Fileset metadata:
```
// Dataframe API
val df = sparkSession
  .read
  .format("fileset")
  .load("gvfs://fileset/fileset_catalog/tmp/test_fileset")
  .where("date=20241021")

df.write()
    .format("fileset")
    .save("gvfs://fileset/fileset_catalog/tmp/test_fileset/date=20241022");

// SQL API
SELECT * FROM `fileset`.`gvfs://fileset/fileset_catalog/tmp/test_fileset` WHERE date = 20241022

INSERT OVERWRITE DIRECTORY 'gvfs://fileset/fileset_catalog/tmp/test_fileset/date=20241023' USING fileset SELECT * FROM test_table;
```
At the same time, since we have a unified read and write data source, users can control when reading data, serializing and deserializing it, and writing it to files. One production case in Xiaomi it that we need to serialize and deserialize it through thrift class when reading and writing sequence files.

**Comment by xloya** [comment]
@jerryshao @shaofengshi @FANNG1 @coolderli What do you think? Please share your opinion when you have time, thanks.

**Comment by coolderli** [comment]
I think this is a very meaningful feature. We can better analyze the fileset while preserving its file semantics. By binding the schema, we can skip the inference stage of the schema for faster analysis.

But I think there are some semantic differences between `fileset.path` and `parquet.path`. The `fileset` is not a specific file format.

**Comment by iodone** [comment]
I think that `fileset` does not have a clear data format definition. It is an abstract concept and cannot be used to specifically implement a datasource. Specific concepts such as` ImageFileset` and `TensorFileset` seem to be implementable.

**Comment by xloya** [comment]
@coolderli @iodone So it's better to bind the specific file format with fileset to make the data source definition clearer? Such as format('sequence-fileset'), format('tfrecord-fileset'). If so, I think it is feasible, but as more file formats are supported, our definitions may also increase.

**Comment by FANNG1** [comment]
It's useful to add this feature, but I have reservations about using the name `fileset` as it not a format.

**Comment by coolderli** [comment]
@xloya @FANNG1 
What about `G-sequence`, `G-tfrecord`, `G-parquet`? Using 
```
select * from `G-sequence`.`gvfs://`
```
to read the data.

There is something I want to be clear. Why we need `G-parquet` not a `parquet`. The `G-parquet` will bind to a table schema and `parquet` will infer schema. Users use this table schema to write data, so we can ensure data compatibility because we assume that the table schema evolves correctly.
Another question is, if we have a table schema, why not write to a table? We need a file format, not a table. For example, in the context of machine learning, we always use a file format, not a table format.

**Comment by xloya** [comment]
> It's useful to add this feature, but I have reservations about using the name `fileset` as it not a format.

So what do you think of this name, we combine the actual file format name with fileset, so that we can distinguish it from the default implementation of other file formats and identify the file format used by Fileset:
```
// Dataframe API
val df = sparkSession
  .read
  **.format("tfrecord-fileset")**
  .load("gvfs://fileset/fileset_catalog/tmp/test_fileset")
  .where("date=20241021")

// SQL API
SELECT * FROM **`tfrecord-fileset`**.`gvfs://fileset/fileset_catalog/tmp/test_fileset` WHERE date = 20241022
```

**Comment by xloya** [comment]
> @xloya @FANNG1 What about `G-sequence`, `G-tfrecord`, `G-parquet`? Using
> 
> ```
> select * from `G-sequence`.`gvfs://`
> ```
> 
> to read the data.
> 
> There is something I want to be clear. Why we need `G-parquet` not a `parquet`. The `G-parquet` will bind to a table schema and `parquet` will infer schema. Users use this table schema to write data, so we can ensure data compatibility because we assume that the table schema evolves correctly. Another question is, if we have a table schema, why not write to a table? We need a file format, not a table. For example, in the context of machine learning, we always use a file format, not a table format.

Yes, I think it is necessary to distinguish from the default file format implementation supported by Spark. Users can of course continue to use Spark's default implementation, but in our data source, we can provide enhanced capabilities. As for the naming of G-parquet, I think it is a little unclear, because currently we only provide it for fileset, and I am not sure whether it will be used for other resources in the future. If not, we'd better bind it to fileset in naming.

**Comment by jerryshao** [comment]
>If users want to read a specific file format that is not officially supported by Spark under Fileset through SQL, there is currently no way.
If the file format does not support schema inference, additional schema metadata is required.

Based on the description, is it more like a file format problem, not a fileset problem? Current fileset addresses the file path and read/write problem, but it doesn't know how to explain the IO stream, so that's why we need to use different "format" to explain the IO stream. 

The problem here is more like adding custom "format" support, am I right? If so, what I'm thinking is that:

1. Spark can support custom file formats other than built-in ones using DSv2, for example, like sequence file, so we can add customized DSv2 support to support different formats.
2. Some file formats are not self-explaining, where to define and store the schema is a question.
3. Fileset is more like a concept of "container", it may have different files with different formats in one fileset, if we want to store "schema" in fileset and use it, we have to make sure that all files follow the same schema/format, otherwise it will be failed to read.

So I was thinking of building an abstract on top of fileset, like @iodone mentioned, `Dataset` with "schemas", and engines can directly read/write this "Dataset" by explicitly/implicitly inferring the "schema", this seems make more sense compared to directly use "fileset" with schema enforced.

**Comment by xloya** [comment]
> > If users want to read a specific file format that is not officially supported by Spark under Fileset through SQL, there is currently no way.
> > If the file format does not support schema inference, additional schema metadata is required.
> 
> Based on the description, is it more like a file format problem, not a fileset problem? Current fileset addresses the file path and read/write problem, but it doesn't know how to explain the IO stream, so that's why we need to use different "format" to explain the IO stream.
> 
> The problem here is more like adding custom "format" support, am I right? If so, what I'm thinking is that:
> 
> 1. Spark can support custom file formats other than built-in ones using DSv2, for example, like sequence file, so we can add customized DSv2 support to support different formats.
> 2. Some file formats are not self-explaining, where to define and store the schema is a question.
> 3. Fileset is more like a concept of "container", it may have different files with different formats in one fileset, if we want to store "schema" in fileset and use it, we have to make sure that all files follow the same schema/format, otherwise it will be failed to read.
> 
> So I was thinking of building an abstract on top of fileset, like @iodone mentioned, `Dataset` with "schemas", and engines can directly read/write this "Dataset" by explicitly/implicitly inferring the "schema", this seems make more sense compared to directly use "fileset" with schema enforced.

Using `Datasets` to unified the concept sounds like a good idea. If I understand correctly, `Datasets` here can be bound to specific schema / file format / serialization infos, so that we can implement different forms of API based on this concept to read the resource types it supports. The API implementation of Datasets here depends entirely on the specific resource type and engine interface. For example, if we support Schema on Read for Fileset in Spark, we can implement File data source based on Datasets to do this; If we support Schema on Read for Table in Spark, we can implement Table data source based on Datasets to do this. I don't know if I understand this correctly.

**Comment by jerryshao** [comment]
Yes, that's what I was thinking about, but it's still my rough thoughts, maybe we can start some basic design to see how it can work.

**Comment by coolderli** [comment]
It is difficult to define the relationship between fileset, dataset, and model. If we introduce a new dataset catalog and model catalog, it may be a bit confusing for users. In addition, datasets or models may only be applicable to Python APIs in machine learning scenarios.

I am more inclined to provide a dataset and model API based on the fileset. We can record more metadata information in fileset, such as schema. And when we use the dataset api, we can use nore metadata information.
```python
# using gvfs api to read the fileset
gvfs.open('gvfs://xxx')
```

``` python
# using dataset api to read the fileset
dataset = datasets.IterableDataset.("catalog.schema.fileset", version='1')
```

**Comment by iodone** [comment]
> It is difficult to define the relationship between fileset, dataset, and model. If we introduce a new dataset catalog and model catalog, it may be a bit confusing for users. In addition, datasets or models may only be applicable to Python APIs in machine learning scenarios.
> 
> I am more inclined to provide a dataset and model API based on the fileset. We can record more metadata information in fileset, such as schema. And when we use the dataset api, we can use nore metadata information.
> 
> ```python
> # using gvfs api to read the fileset
> gvfs.open('gvfs://xxx')
> ```
> 
> ```python
> # using dataset api to read the fileset
> dataset = datasets.IterableDataset.("catalog.schema.fileset", version='1')
> ```

Referring to the Python API of the `Lance` dataset, do we consider implementing lance's catalog based on fileset? Fileset is an open definition without any schema constraints, unless we build a fileset data format ourselves to support unstructured data (supporting image/tfrecord).

**Comment by jerryshao** [comment]
> Referring to the Python API of the Lance dataset, do we consider implementing lance's catalog based on fileset?

Hi @iodone would you please point me out where I can see the lance's catalog details?

**Comment by coolderli** [comment]
> > Referring to the Python API of the Lance dataset, do we consider implementing lance's catalog based on fileset?
> 
> Hi @iodone would you please point me out where I can see the Lance's catalog details?

@jerryshao I'm not sure if we need a lance catalog. But I do want a unity API to read the lancedb.
https://lancedb.github.io/lance/read_and_write.html
